### PR TITLE
[Tests] Add linter for circular dependencies and start fixing them

### PR DESCRIPTION
**Pull request by random-zebra**
Backports from upstream a shell script which uses the the python script `circular-dependencies.py` (in devtools) and gives a convenient feedback about any dependency introduced and/or removed. 
This will be run by GA on any PR (as part of `lint-all.sh`)

Also, start nuking the first easy circular dependencies caused by unneeded header imports.

### CODE REVIEW

File: src/sapling/noteencryption.h [hunk 1]
```diff
@@ -13,10 +13,10 @@ See the Zcash protocol specification for more information.
 
 #include "uint256.h"
 
-#include "sapling/address.h"
 #include "sapling/sapling.h"
 
 #include <array>
+#include <boost/optional.hpp>
```
Comment (random-zebra) [inline src/sapling/noteencryption.h:19]: note to self: conflict with #2642

File: test/lint/lint-circular-dependencies.sh [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    if [[ ${IS_PRESENT_EXPECTED_CIRC} == 0 ]]; then
+        echo "Good job! The circular dependency \"${EXPECTED_CIRC}\" is no longer present."
+        echo "Please remove it from EXPECTED_CIRCULAR_DEPENDENCIES in $0"
+        echo "to make sure this circular dependency is not accidentally reintroduced."
+        echo
+        EXIT_CODE=1
+    fi
+done
+
+exit ${EXIT_CODE}
```
Comment (furszy) [inline test/lint/lint-circular-dependencies.sh:134]: tiny nit: jump line at the end of the file.

**Review by furszy — APPROVED** [review]
very nice :). Now that this problems are visible, anyone can work on them 🚜.
ACK 95d401146166ee04cf44d1a4affae7c4249c7dea

File: test/lint/lint-circular-dependencies.sh (hunk 2 above)
Comment (random-zebra) [reply test/lint/lint-circular-dependencies.sh:134]: :+1: will add the jump line in either #2643, #2644 or #2646 (which are all editing this file, and will all need to be rebased when this PR is merged).

**Review by Fuzzbawls — APPROVED** [review]
ACK 95d401146166ee04cf44d1a4affae7c4249c7dea

**Pull request merged**